### PR TITLE
fix logout crash and provide completion block to notify completion of…

### PR DESCRIPTION
**Pull request by thuyBox**
… background session's setup,

and background tasks can be provided instead of blocking call on background task retrieval

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @thuyBox has signed the CLA. Thanks for the pull request!

File: BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m [hunk 1]
```diff
@@ -174,7 +174,7 @@ - (BOOL)cacheBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(
         }
         return NO;
     }
-    NSData *data = error == nil ? nil : [NSKeyedArchiver archivedDataWithRootObject:taskError];
+    NSData *data = taskError == nil ? nil : [NSKeyedArchiver archivedDataWithRootObject:taskError];
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:177]: oops

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.h [hunk 2]
```diff
@@ -192,9 +197,15 @@ totalBytesExpectedToSend:(int64_t)totalBytesExpectedToSend;
...
+/**
+ * Cancel and clean up all background session tasks associated with this user
+ */
 - (void)cancelAndCleanUpBackgroundSessionTasksForUserId:(NSString *)userId error:(NSError **)outError;
 
-- (void)cacheResumeData:resumeData forUserId:(NSString *)userId associateId:(NSString *)associateId;
+/**
+ * Cache resumeData for background download session task of this userId and associateId
+ */
+- (BOOL)cacheResumeData:resumeData forUserId:(NSString *)userId associateId:(NSString *)associateId;
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionManager.h]: What type is resumeData? This should be cacheResumeData:(NSData *)resumeData forUserID:(... right?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m [hunk 3]
```diff
@@ -125,19 +122,11 @@ - (id)init
         _backgroundSessionIdToSession = [NSMutableDictionary new];
         [self createDefaultSession];
         [self createProgressSession];
-        self.finishSettingUpBackgroundSession = NO;
-        backgroundSessionSetupSemphore = dispatch_semaphore_create(0);
+        _finishSettingUpBackgroundSession = NO;
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m]: Were we not updating the name to something like didFinishSettingUp or hasFinishedSettingUp?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m [hunk 4]
```diff
@@ -305,7 +315,7 @@ - (NSURLSession *)backgroundSessionForId:(NSString *)backgroundSessionId
 
 - (void)pendingBackgroundDownloadUploadSessionTasks:(void (^)(NSArray<NSURLSessionUploadTask *> * uploadTasks, NSArray<NSURLSessionDownloadTask *> * downloadTasks))completion
 {
-    if (completion != nil) {
+    if (self.supportBackgroundSessionTasks == YES && completion != nil) {
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m]: if there is a completion shouldn't there be an else where we return the 'not supported' error? That's what we've done elsewhere...

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m [hunk 5]
```diff
@@ -400,12 +410,17 @@ - (NSURLSessionUploadTask *)backgroundUploadTaskWithRequest:(NSURLRequest *)requ
...
+
+- (BOOL)supportBackgroundSessionTasks
+{
+    return self.backgroundSession != nil && self.cacheClient != nil;
 }
 
 - (BOOL)cleanUpSessionTaskInfoGivenUserId:(NSString *)userId associateId:(NSString *)associateId error:(NSError **)outError
 {
-    if (userId == nil || associateId == nil) {
+    if (self.supportBackgroundSessionTasks == NO || userId == nil || associateId == nil) {
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m]: shouldn't cleanup happen no matter what?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m [hunk 6]
```diff
@@ -448,7 +463,7 @@ - (void)cancelAndCleanUpBackgroundSessionTasksForUserId:(NSString *)userId error
 {
     //FIXME: make sure set up is completed, and prevents any new tasks created for this userId while we clean it up
 
-    if (userId != nil) {
+    if (self.supportBackgroundSessionTasks == YES && userId != nil) {
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m:473]: shouldn't cleanup happen no matter what?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m [hunk 7]
```diff
@@ -536,7 +551,14 @@ - (BOXBackgroundSessionIdAndTask *)existingBackgroundSessionTaskGivenUserId:(NSS
 //      re-create the download task with resume data and return that
 - (NSURLSessionTask *)backgroundTaskWithRequest:(NSURLRequest *)request orUploadFromFile:(NSURL *)uploadFromFileURL taskDelegate:(id <BOXURLSessionTaskDelegate>)taskDelegate userId:(NSString *)userId associateId:(NSString *)associateId error:(NSError **)outError
 {
-    [self waitForBackgroundSessionSetUpToFinish];
+    if (self.finishSettingUpBackgroundSession == NO) {
+        //trying to get a background session task before setting up background session completes
+        //will fail this method
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m]: how likely is this to happen?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.h (hunk 2 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionManager.h]: updated. :( didn't have a problem running.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m (hunk 3 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m (hunk 4 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m]: done

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m (hunk 5 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m]: this only applies for background session task, will update the title accordingly

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m (hunk 6 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m:473]: this only applies if we support background session tasks, if user never sets up to support background tasks, there will be nothing to clean.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m (hunk 7 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m]: this is unlikely. The app is expected to wait for completionBlock of the one time set up of background session to request for background tasks.

**Review by boxtcarpel — APPROVED** [review]

**Pull request merged**